[3.x] Tooltips for exported variables

Listwon opened the pull request:
Backward compatible solution to add custom tooltips for exported variables.
Uses comments syntax, so it won't throw errors on previous versions. Tooltips are parsed only in editor. Supports BBCode.
``` gdscript
## [b]Tooltips[/b] for exported variables. [br][img]icon.png[/img] [br]With [color=green]BBCode[/color] support. [br]Enjoy!
export var with_tooltip = 2
```
![tooltips1](https://user-images.githubusercontent.com/1554127/73392683-8df26c00-42da-11ea-9fc5-f6287db5b410.jpg)
![tooltips2](https://user-images.githubusercontent.com/1554127/73392659-8632c780-42da-11ea-9d96-a67a7f9f3a83.jpg)

Related issue https://github.com/godotengine/godot/issues/6204

Listwon committed:
Tooltips for exported variables

Calinou commented:
This looks great! I'm kind of amazed it was that simple to implement.

Still, we may want to look into adding a basic Markdown -> BBCode formatter. Writing BBCode in comments makes it less readable, whereas Markdown source is still easy to read.

Geequlim commented:
I'm not sure it is a good idea to add it to PropertyInfo.
As it is not used by properties which are not exported. And it is useless for release version of our games.

https://github.com/godotengine/godot/blob/6fcb58f40dae8228074591e3262d5db97db3e2d7/modules/gdscript/language_server/gdscript_extend_parser.h#L94

In fact you can get the documentations of the exported variables from the LSP extended parser with multi-lines supported.

Listwon commented:
@Geequlim 

> I'm not sure it is a good idea to add it to PropertyInfo.

Yeah, I guess it would be better to wrap this in TOOLS_ENABLED block as well.

> In fact you can get the documentations of the exported variables from the LSP extended parser with multi-lines supported.

I get it, but the problem solved here is different - to provide the tool to show helpful info in editor, especially useful for non-programmers. As I tried your parser with VS Code it gets the whole comment above the variable. You can't explicitly mark something as documentation, right?

aaronfranke commented:
Could this be expanded to be a general system for inline documentation, similar to the `/// <summary>` XML comments in C#? There's an issue somewhere about generating in-editor documentation for user-created scripts.

Calinou commented:
@aaronfranke The proposal you're referring to is https://github.com/godotengine/godot-proposals/issues/177.

ThakeeNathees commented on modules/gdscript/gdscript_tokenizer.h:
```diff
@@ -142,6 +142,7 @@ class GDScriptTokenizer {
 		TK_ERROR,
 		TK_EOF,
 		TK_CURSOR, //used for code completion
+		TK_TOOLTIP,
```
creating a new token would break the parser unless every corner cases were handled.
ex: if `TK_TOOLTIP` inside a function the parser complains `Expected end of statement after expression, got ...` inside `_parse_block()`
and we check the next and previous tokens for parsing so we need to tell the parser that if it's a tool tip token just skip it.

willnationsdev commented:
Is this something that @vnen should be considering as another type of GDScript annotation?

vnen commented:
Yeah, this should be added as an annotation.

However, @ThakeeNathees is working on showing GDScript documentation

ShivamMukherjee commented:
Eagerly waiting for this to get merged and usable!

Calinou commented:
@ShivamMukherjee This pull request will most likely conflict with https://github.com/godotengine/godot/pull/39359 and the GDScript rewrite, so I doubt it'll be merged.

ThakeeNathees commented:
@Calinou should we consider export variable doc comments as tool tips ?

ShivamMukherjee commented:
I'll be Waiting for Godot then! As long as I get to use tooltips like a sane person again.

willnationsdev commented:
@ThakeeNathees only if the export variable's doc comments are also rendered as editor tooltips. Has that been confirmed by @vnen?

ThakeeNathees commented:
@willnationsdev No, I'm just suggesting that to use it as a tooltip yet no decision has made.

ThakeeNathees commented:
@willnationsdev , @vnen agreed to use the documentation of export variable to use it as tooltip, 

here is a working example 
![Capture](https://user-images.githubusercontent.com/41085900/89592922-7688d280-d86b-11ea-9f92-6c90509cf4d3.JPG)
![Screenshot (28)](https://user-images.githubusercontent.com/41085900/89592931-7b4d8680-d86b-11ea-978c-fc993c2ffe44.png)

Calinou commented:
This feature is now available in `master` (what will become 4.0 in the future) thanks to https://github.com/godotengine/godot/pull/41095.

Should we aim to support this in `3.2`? We should aim to use a compatible syntax so that projects can be upgraded smoothly without having to update comments.

cc @vnen

Calinou commented on misc/dist/linux/godot.6:
```diff
@@ -1,4 +1,4 @@
-.TH GODOT "6" "January 2020" "godot 3.2" "Games"
+.TH GODOT "6" "January 2020" "godot 4.0" "Games"
```
Unnecessary change (due to rebase error?), please revert it.

Calinou commented on misc/dist/osx_tools.app/Contents/Info.plist:
```diff
@@ -19,11 +19,11 @@
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>3.2</string>
+	<string>4.0</string>
```
Unnecessary change (due to rebase error?), please revert it.

Calinou commented on misc/dist/osx_tools.app/Contents/Info.plist:
```diff
@@ -19,11 +19,11 @@
 	<string>godot</string>
 	<key>CFBundleVersion</key>
-	<string>3.2</string>
+	<string>4.0</string>
```
Unnecessary change (due to rebase error?), please revert it.

Calinou commented on editor/inspector_dock.cpp:
```diff
@@ -584,6 +584,7 @@ InspectorDock::InspectorDock(EditorNode *p_editor, EditorData &p_editor_data) {
 	add_child(warning);
 	warning->set_text(TTR("Changes may be lost!"));
 	warning->set_icon(get_icon("NodeWarning", "EditorIcons"));
+	warning->set_clip_text(true);
```
Out of curiosity, is there a reason this was set here? Or does this fix an unrelated problem?

Calinou commented on modules/gdscript/gdscript_tokenizer.cpp:
```diff
@@ -135,7 +135,8 @@ const char *GDScriptTokenizer::token_names[TK_MAX] = {
 	"EOF",
-	"Cursor"
+	"Cursor",
+	"##" //tooltip
```
Add a trailing comma for easier refactoring, and tweak the comment style to match what we use now:

```suggestion
	"##", // Tooltip comment for exported variables.
```

Calinou commented on version.py:
```diff
@@ -1,9 +1,9 @@
 name = "Godot Engine"
-major = 3
-minor = 2
+major = 4
+minor = 0
```
Unnecessary change (due to rebase error?), please revert it.

Calinou commented on version.py:
```diff
@@ -1,9 +1,9 @@
+minor = 0
 patch = 0
-status = "stable"
+status = "dev"
```
Unnecessary change (due to rebase error?), please revert it.

Calinou reviewed (changes requested):
I think this would be a welcome change for 3.4 since the syntax is forwards-compatible with 4.0. However, https://github.com/godotengine/godot/pull/35716/files#r389275266 needs to be addressed before this can be merged.

Listwon commented:
@Calinou I fixed auto-rebase issues and the comment with trailing comma in the same line. I've also added the solution for multiline tooltip comments, which you asked about at the very beginning and I missed this earlier.

DaGamingWolf commented:
man i would really like this feature. i'm designing an interface for our writers to use without touching much code and tooltips would be massively helpful.

KoBeWi commented:
Apparently this implementation works with built-in scripts, unlike the 4.0 one. However, it breaks compatibility, due to the token issue mentioned before. I tend to put `##` in comments for stuff I want to check layer, so it's easier to find. If this was merged, all scripts using `##` "incorrectly" will break.

Listwon commented:
@KoBeWi could you please test those corner cases on newer version of my PR? I found some issues in the release build too, so it should be covered now. Right now it will generate token only if `##` comment is just before export variable.

KoBeWi reviewed (approved):
Looks ok feature-wise, but GDScript changes need review.

Also we should probably document this somewhere, including the BBCode syntax (although I assume it's the same as in RTL). Especially the multiline comments might be confusing, as you need to use `[br]`.

Vivraan commented:
How difficult is it to make a markdown or even standard XML comment parser?

Calinou commented:
> How difficult is it to make a markdown or even standard XML comment parser?

I wouldn't add Markdown syntax here, as BBCode is what we went for in 4.0.

geekley commented:
Does anyone know if tooltips from documentation (e.g. XMLDoc) are available for C# `[Export]` members in 4.0? I tried a simple test and it doesn't seem like it's supported (yet?).

If not, where is an issue/PR I can watch to track this? Sorry if this is the wrong place to ask. I tried to find a specific issue for C#, but I only found this. Should I open a new issue?

For example, something like this:
```cs
///<summary>Tooltip blah blah </summary>
[Export] string myProp;
```
Or even something like this, if you want to be more flexible?
```cs
///<tooltip><summary>Tooltip blah blah </summary></tooltip>
///<remarks>Not in tooltip.</remarks>
[Export] string myProp;
```

Calinou commented:
@geekley It's only supported in GDScript in 4.0. This is not the place to discuss C# support though; please open a [proposal](https://github.com/godotengine/godot-proposals) for that.

geekley commented:
Thanks Calinou! I was going to, but turns out there is a proposal already: https://github.com/godotengine/godot-proposals/discussions/5675
Linking here in case people want to upvote it.

turbohz commented:
There's also properties exported via `_get_property_list()`.

Looks like something that should be tightly related, but I'm not familiar with the code.

I wonder if implementing this for those should better be done in another PR, 
or could just be done in one go.

aaronfranke commented:
@Listwon This is a nice feature, and the `##` syntax is forwards-compatible with the documentation system in Godot 4.x. Can you rebase this PR and test if it still works?

Above @vnen and @willnationsdev discussed annotations, but Godot 4.x uses the `##` syntax, so the syntax in this PR is correct to be consistent with Godot 4.x.

lawnjelly commented:
There's been an effort to rebase this in #120429.

Just to get the original reviewers involved, is this PR considered ok to move ahead with, are there any outstanding issues to resolve (especially breaking parsers)?

I do certainly worry that the `##` format could potentially be breaking in that there may be existing comments in people's games that contain `##`, and maintaining backward compatibility is paramount in 3.x. If there were existing comments like this, presumably the risk is in unwanted tooltips, or could this in fact break the new parsing and cause crashes or similar?

But if 4.x already uses this, there is some logic to maintaining the same method.

Options might include a project setting to turn this behaviour on and off, but then that would cause problems with discoverability if defaulting to off for compatibility. But we could e.g. default to on, and thus provide an easy path for upgrade. 

Another option is to actually use a more unique token to start the tooltip, e.g.
`##tooltip## BODY`
or indeed the annotion type
`@tooltip BODY`
or some variation.

Forward compatibility is less of an issue now for 3.7 as most people upgrading will have already done it (but keeping compat would be nice).

It does seem quite desired but may open up knock on bugs, and we want to minimise these / ensure someone is keen to maintain this new code and fix them.

aaronfranke commented:
@lawnjelly I don't think "helpful tooltips now appear when they didn't before" should be even slightly considered to be a breaking change. If that is the only concern... well, it shouldn't be a concern at all, so I'd say it's good to go.

Listwon commented:
In 2022 my PR was waiting for a review and it seems it was not accepted (no reviews since then). I don't feel it's ok to copy it 1:1 (instead of cherry-picking which would preserve me as an author). I didn't have time to try to rebase it and forgot to do so later, sorry. It still works but still requires a review and maybe a set of tests to check if it doesn't break any edge cases, which will take time and effort.
<img width="416" height="237" alt="image" src="https://github.com/user-attachments/assets/01592d0c-4d7b-41f5-a6f7-8a28e0a9920e" />

lawnjelly commented:
> I don't feel it's ok to copy it 1:1 (instead of cherry-picking which would preserve me as an author).

Yes, fully agree. I was intending to rebase / cherry pick yours (or ensure you got credit in the revised PR if there were later contributions).

> I don't think "helpful tooltips now appear when they didn't before" should be even slightly considered to be a breaking change. If that is the only concern... well, it shouldn't be a concern at all, so I'd say it's good to go.

Agreed, but that's what I'm asking really.

Are there risks of `##` feeding existing comments into new parsing code for the tooltips, and choking that parser?
e.g.

`## he898y23ry fwejfwe? fijwefij23888`

or is the tooltip parsing code fairly robust already, and able to cope with any existing garbage comments in users projects?
